main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+    case ResolvedAst.Pattern.Int64(lit, loc) => KindedAst.Pattern.Int64(lit, loc).toSuccess
+    case ResolvedAst.Pattern.BigInt(lit, loc) => KindedAst.Pattern.BigInt(lit, loc).toSuccess
+    case ResolvedAst.Pattern.Str(lit, loc) => KindedAst.Pattern.Str(lit, loc).toSuccess
+    case ResolvedAst.Pattern.Tag(sym, tag, pat0, loc) =>
+      for {
+        pat <- ascribePattern(pat0, kenv, root)
+      } yield KindedAst.Pattern.Tag(sym, tag, pat, Type.freshVar(Kind.Star), loc)
+    case ResolvedAst.Pattern.Tuple(elms0, loc) =>
+      for {
+        elms <- Validation.traverse(elms0)(ascribePattern(_, kenv, root))
+      } yield KindedAst.Pattern.Tuple(elms, loc)
+    case ResolvedAst.Pattern.Array(elms0, loc) =>
+      for {
+        elms <- Validation.traverse(elms0)(ascribePattern(_, kenv, root))
+      } yield KindedAst.Pattern.Array(elms, Type.freshVar(Kind.Star), loc)
+    case ResolvedAst.Pattern.ArrayTailSpread(elms0, sym, loc) =>
+      for {
+        elms <- Validation.traverse(elms0)(ascribePattern(_, kenv, root))
+      } yield KindedAst.Pattern.ArrayTailSpread(elms, sym, Type.freshVar(Kind.Star), loc)
+    case ResolvedAst.Pattern.ArrayHeadSpread(sym, elms0, loc) =>
+      for {
+        elms <- Validation.traverse(elms0)(ascribePattern(_, kenv, root))
+      } yield KindedAst.Pattern.ArrayHeadSpread(sym, elms, Type.freshVar(Kind.Star), loc)
+  }
+
+
+  /**
+    * Performs kinding on the given choice pattern under the given kind environment.
+    */
+  private def ascribeChoicePattern(pat0: ResolvedAst.ChoicePattern, kinds: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.ChoicePattern, KindError] = pat0 match {
+    case ResolvedAst.ChoicePattern.Wild(loc) => KindedAst.ChoicePattern.Wild(loc).toSuccess
+    case ResolvedAst.ChoicePattern.Absent(loc) => KindedAst.ChoicePattern.Absent(loc).toSuccess
+    case ResolvedAst.ChoicePattern.Present(sym, loc) => KindedAst.ChoicePattern.Present(sym, Type.freshVar(Kind.Star), loc).toSuccess
+  }
+
+  /**
+    * Performs kinding on the given constraint under the given kind environment.
+    */
+  private def ascribeConstraint(constraint0: ResolvedAst.Constraint, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Constraint, KindError] = constraint0 match {
+    case ResolvedAst.Constraint(cparams0, head0, body0, loc) =>
+      for {
+        cparams <- Validation.traverse(cparams0)(ascribeCparam(_, kenv, root))
+        head <- ascribeHeadPredicate(head0, kenv, root)
+        body <- Validation.traverse(body0)(ascribeBodyPredicate(_, kenv, root))
+      } yield KindedAst.Constraint(cparams, head, body, loc)
+  }
+
+  /**
+    * Performs kinding on the given constraint param under the given kind environment.
+    */
+  private def ascribeCparam(cparam0: ResolvedAst.ConstraintParam, kenv: KindEnv, root: ResolvedAst.Root): Validation[KindedAst.ConstraintParam, KindError] = cparam0 match {
```

Comment:
In function names we dont shorten to fparam or cparam. Its usually `visitFormalParam` etc.